src/main/java/com/icthh/xm/ms/entity/service/impl/FunctionServiceImpl.java

```diff
@@ -79,40 +83,73 @@ public FunctionContext execute(final String functionKey,
 
         Map<String, Object> vInput = CustomCollectionUtils.emptyIfNull(functionInput);
 
+        dynamicPermissionCheckService.checkContextPermission(DynamicPermissionCheckService.FeatureContext.FUNCTION,
+            "XMENTITY.FUNCTION.EXECUTE", functionKey);
+
         // get type key
-        XmEntityIdKeyTypeKey projection = xmEntityService.getXmEntityIdKeyTypeKey(idOrKey);
-        String xmEntityTypeKey = projection.getTypeKey();
+        XmEntityStateProjection projection = xmEntityService.findStateProjectionById(idOrKey).orElseThrow(
+            () -> new EntityNotFoundException("XmEntity with idOrKey [" + idOrKey + "] not found")
+        );
 
         // validate that current XmEntity has function
-        FunctionSpec functionSpec = xmEntitySpecService.findFunction(xmEntityTypeKey, functionKey).orElseThrow(
-            () -> new IllegalArgumentException("Function not found for entity type key " + xmEntityTypeKey
-                + " and function key: " + functionKey)
-        );
+        FunctionSpec functionSpec = findFunctionSpec(functionKey, projection);
+
+        isCallAllowedByState(functionSpec, projection).orElseThrow(
+            () -> new IllegalStateException("Function call forbidden for current state"));
 
         // execute function
-        Map<String, Object> data = functionExecutorService.execute(functionKey, idOrKey, xmEntityTypeKey, vInput);
+        Map<String, Object> data = functionExecutorService.execute(functionKey, idOrKey, projection.getTypeKey(), vInput);
+
+        return processFunctionResult(functionKey, idOrKey, data, functionSpec);
 
-        // save result in FunctionContext
-        if (functionSpec.getSaveFunctionContext()) {
-            return saveResult(functionKey, idOrKey, data, functionSpec);
-        } else {
-            return toFunctionContext(functionKey, idOrKey, data, functionSpec);
-        }
     }
 
     /**
-     * Execute any function.
-     *
-     * @param functionKey function key
-     * @param idOrKey     XmEntity id or key (can be {@code null})
-     * @return function execution result context
+     * Validates, if current entity state is one of allowed states
+     * @param functionSpec - functionSpec
+     * @param projection - entity projection
+     * @return Optional.of(Boolean.TRUE) if allowed or Optional.Empty if not
      */
-    private FunctionContext saveResult(String functionKey,
-                                       IdOrKey idOrKey,
-                                       Map<String, Object> data,
-                                       FunctionSpec functionSpec) {
+    protected Optional<Boolean> isCallAllowedByState(FunctionSpec functionSpec, XmEntityStateProjection projection) {
+        List<String> allowedStates = CustomCollectionUtils.nullSafe(functionSpec.getAllowedStateKeys());
```

Comment:
it's field already used for hide function on ui
but function call from widgets and api
please help me to understand how we keep backward compatibility

Reply:
UI check is not guarantee than function is not executed in this state. This should be also checked on BE. As standard behavior  it will help not to write this code in LEP

Reply:
also, may be better approach will be

UI only configuration (adjustment) should be in webapp-spec (like error mapping) and other UI configs
If something is in entity spec, this should be evaluated as overall be+fe behavior
what is the point to hide something on entity card, but to allow to call in other places?

Reply:
**backward compatibility** - good point, _featureFlag_ should be added.

Reply:
@sergeysenja1992 @major13ua @ishkurko 
i think we need to leave `allowedStateKeys` execution logic without changes for now.

But, agree that this flag is ambiguous because that name does not clearly defines that is influence only UI.
It shoyld be renamed to probably `showInStateKeys` (intorduce alias with same logic, then deprecate allowedStateKeys and then remove).


meanwhile lets consider several ways of resolution for this problem:

1) Introduce separate container executeInStateKeys to be checked on BE:
showInStateKeys | allowedStateKeys:
    - STATE1
    - STATE2
executeInStateKeys:
    - STATE1
    - STATE2
    - STATE3
(Note, we will need some grace period while both flags showInStateKeys and allowedStateKeys will be supported for smooth migration)

2) Implement allowedStateKeys logic on UI spec and start checking on BE:
--- public-settings.yml // UI spec
allowedStateKeys:
    - STATE1
    - STATE2
--- xmentityspec.yml:
allowedStateKeys:
    - STATE1
    - STATE2
    - STATE3

(Note, in this case Function definition will be splitted between two files)

3) any other ways?

Reply:
in this case we have no other way, to add featureFlag
if we added featureFlags in other cases, why not add in this?